jmclient/jmclient/auth.py

```diff
@@ -0,0 +1,94 @@
+import datetime
+import os
+from typing import Optional
+
+import jwt
+
+from jmbase.support import bintohex
+
+
+class InvalidScopeError(Exception):
+    pass
+
+
+class JMSessionValidity:
+    access = datetime.timedelta(minutes=30)
+    refresh = datetime.timedelta(hours=4)
+
+
+class JMTokenSignatureKey:
+    algorithm = "HS256"
+
+    def __init__(self):
+        self.access = self.get_random_key()
+        self.refresh = self.get_random_key()
+
+    @staticmethod
+    def get_random_key(size: int = 16) -> str:
+        """Create a random key has an hexadecimal string."""
+        return bintohex(os.urandom(size))
+
+    def reset(self, refresh_token_only: bool):
+        """Invalidate previously issued token(s) by creating new signature key(s)."""
+        self.refresh = self.get_random_key()
+        if not refresh_token_only:
+            self.access = self.get_random_key()
+
+
+class JMTokenAuthority:
+    """Manage authorization tokens."""
+
+    validity = JMSessionValidity()
+
+    def __init__(self, wallet_name: Optional[str] = None):
+        self.signature = JMTokenSignatureKey()
+        self.wallet_name = wallet_name
+
+    def verify(
+        self,
+        token: str,
+        scope: str = "walletrpc",
+        *,
+        is_refresh: bool = False,
+        verify_exp: bool = True,
+    ):
+        """Verify JWT token.
+
+        Token must have a valid signature and its scope must contain both scopes in
+        arguments and wallet_name property.
+        """
+        token_type = "refresh" if is_refresh else "access"
+        claims = jwt.decode(
+            token,
+            getattr(self.signature, token_type),
+            algorithms=self.signature.algorithm,
+            options={"verify_exp": verify_exp},
+            leeway=10,
+        )
+        token_claims = set(claims.get("scope", []).split())
+        if not set(scope.split()) | {self.wallet_name} <= token_claims:
+            raise InvalidScopeError
+
+    def _issue(self, scope: str, token_type: str) -> str:
+        return jwt.encode(
+            {
+                "exp": datetime.datetime.utcnow() + getattr(self.validity, token_type),
+                "scope": f"{scope} {self.wallet_name}",
+            },
+            getattr(self.signature, token_type),
+            algorithm=self.signature.algorithm,
+        )
+
+    def issue(self, scope: str = "walletrpc") -> dict:
```

Comment:
Nit: perhaps default string should include 'joinmarket' in it, to specify scope better.

Reply:
I wondered as well. I ended up with the "shortened version" considering scope is encrypted and solely readable by `JMTokenAuthority`, it is internal only. We could skip it entirely at this stage but I thought it was good to leave a door open, would we need different authorization level in the future.